### PR TITLE
Introduce temporary variables for calls and ternary expressions

### PR DESCRIPTION
**Pull request by Nargeshdb**

### CODE REVIEW

File: object-construction-checker/src/main/java/org/checkerframework/checker/objectconstruction/ObjectConstructionAnnotatedTypeFactory.java [hunk 1]
```diff
@@ -31,6 +35,7 @@
  */
 public class ObjectConstructionAnnotatedTypeFactory extends CalledMethodsAnnotatedTypeFactory {
 
+  protected BiMap<LocalVariableNode, Node> biMap = HashBiMap.create();
```
Comment (msridhar) [inline object-construction-checker/src/main/java/org/checkerframework/checker/objectconstruction/ObjectConstructionAnnotatedTypeFactory.java:38]: This field needs a better name than `biMap` and some documentation

File: object-construction-checker/src/main/java/org/checkerframework/checker/objectconstruction/ObjectConstructionAnnotatedTypeFactory.java [hunk 2]
```diff
@@ -73,6 +78,9 @@ public void postAnalyze(ControlFlowGraph cfg) {
   List<String> getMustCallValue(Tree tree) {
     MustCallAnnotatedTypeFactory mustCallAnnotatedTypeFactory =
         getTypeFactoryOfSubchecker(MustCallChecker.class);
+    if (mustCallAnnotatedTypeFactory == null) {
```
Comment (msridhar) [inline object-construction-checker/src/main/java/org/checkerframework/checker/objectconstruction/ObjectConstructionAnnotatedTypeFactory.java:81]: How is this change related to temporary variables?  Same for the next change below

File: object-construction-checker/src/main/java/org/checkerframework/checker/objectconstruction/ObjectConstructionTransfer.java [hunk 3]
```diff
@@ -45,6 +55,8 @@
 public class ObjectConstructionTransfer extends CalledMethodsTransfer {
   private final ObjectConstructionAnnotatedTypeFactory atypefactory;
 
+  public final TreeBuilder treeBuilder;
```
Comment (msridhar) [inline object-construction-checker/src/main/java/org/checkerframework/checker/objectconstruction/ObjectConstructionTransfer.java:58]: Should probably be `private`?  Also document what this is for

File: object-construction-checker/src/main/java/org/checkerframework/checker/objectconstruction/ObjectConstructionTransfer.java [hunk 4]
```diff
@@ -74,9 +88,54 @@ public ObjectConstructionTransfer(final CFAnalysis analysis) {
             result.getElseStore(),
             exceptionalStores);
     exceptionalStores = null;
+
+    Node receiver = node.getTarget().getReceiver();
+    addTemporaryVars(result, node);
+    if (atypefactory.biMap.inverse().containsKey(receiver)) {
```
Comment (msridhar) [inline object-construction-checker/src/main/java/org/checkerframework/checker/objectconstruction/ObjectConstructionTransfer.java:94]: To me this code would be clearer if you stored the result of `atypefactory.biMap.inverse().get(receiver)` in a local variable, and then changed the condition to check if that variable is null.  Then you could avoid writing `atypefactory.biMap.inverse()` in two places

File: object-construction-checker/src/main/java/org/checkerframework/checker/objectconstruction/ObjectConstructionTransfer.java [hunk 5]
```diff
@@ -74,9 +88,54 @@ public ObjectConstructionTransfer(final CFAnalysis analysis) {
...
 
+  @Override
+  public TransferResult<CFValue, CFStore> visitObjectCreation(
+      ObjectCreationNode node, TransferInput<CFValue, CFStore> input) {
+    TransferResult<CFValue, CFStore> result = super.visitObjectCreation(node, input);
+    addTemporaryVars(result, node);
+    return result;
+  }
+
+  private void addTemporaryVars(TransferResult<CFValue, CFStore> result, Node node) {
```
Comment (msridhar) [inline object-construction-checker/src/main/java/org/checkerframework/checker/objectconstruction/ObjectConstructionTransfer.java:113]: Please add Javadoc

File: object-construction-checker/src/main/java/org/checkerframework/checker/objectconstruction/ObjectConstructionTransfer.java [hunk 6]
```diff
@@ -74,9 +88,54 @@ public ObjectConstructionTransfer(final CFAnalysis analysis) {
...
+    if (atypefactory.hasMustCall(node.getTree())) {
+      if (atypefactory.hasMustCall(node.getTree())) {
```
Comment (msridhar) [inline object-construction-checker/src/main/java/org/checkerframework/checker/objectconstruction/ObjectConstructionTransfer.java:115]: Redundant with previous line

File: object-construction-checker/src/main/java/org/checkerframework/checker/objectconstruction/ObjectConstructionTransfer.java (hunk 5 above)
Comment (msridhar) [reply object-construction-checker/src/main/java/org/checkerframework/checker/objectconstruction/ObjectConstructionTransfer.java:113]: Also this method needs a better name.  Maybe `updateStoreWithTempVar`?  I think the method either creates or looks up the temp var `t` for `node`, and then updates the store to give `t` the same type as `node`.  Is that correct?

File: object-construction-checker/src/main/java/org/checkerframework/checker/objectconstruction/ObjectConstructionTransfer.java [hunk 7]
```diff
@@ -74,9 +88,54 @@ public ObjectConstructionTransfer(final CFAnalysis analysis) {
...
+        LocalVariableNode localVariableNode;
+        if (!atypefactory.biMap.inverse().containsKey(node)) {
+          VariableTree temp = createTemporaryVar(node);
+          IdentifierTree identifierTree = treeBuilder.buildVariableUse(temp);
+          localVariableNode = new LocalVariableNode(identifierTree);
+          localVariableNode.setInSource(true);
+          atypefactory.biMap.put(localVariableNode, node);
+        } else {
+          localVariableNode = atypefactory.biMap.inverse().get(node);
+        }
```
Comment (msridhar) [inline object-construction-checker/src/main/java/org/checkerframework/checker/objectconstruction/ObjectConstructionTransfer.java:126]: I would extract this code into its own method, `getOrCreateTempVar`

File: object-construction-checker/src/main/java/org/checkerframework/checker/objectconstruction/ObjectConstructionTransfer.java [hunk 8]
```diff
@@ -74,9 +88,54 @@ public ObjectConstructionTransfer(final CFAnalysis analysis) {
             result.getElseStore(),
             exceptionalStores);
     exceptionalStores = null;
+
+    Node receiver = node.getTarget().getReceiver();
+    addTemporaryVars(result, node);
```
Comment (msridhar) [inline object-construction-checker/src/main/java/org/checkerframework/checker/objectconstruction/ObjectConstructionTransfer.java:93]: I'd move this line before the previous one, to keep the code about the `receiver` together.  Then, add some comments indicating what the code involving `receiver` is doing

File: object-construction-checker/src/main/java/org/checkerframework/checker/objectconstruction/ObjectConstructionTransfer.java [hunk 9]
```diff
@@ -167,9 +227,41 @@ public void accumulate(Node node, TransferResult<CFValue, CFStore> result, Strin
...
         }
       }
+      AnnotationMirror newAnno = atypefactory.createAccumulatorAnnotation(valuesAsList);
+      exceptionalStores.values().stream().forEach(s -> s.insertValue(target, newAnno));
     }
-    AnnotationMirror newAnno = atypefactory.createAccumulatorAnnotation(valuesAsList);
-    exceptionalStores.values().stream().forEach(s -> s.insertValue(target, newAnno));
+  }
+
+  private void insertIntoStores(
```
Comment (msridhar) [inline object-construction-checker/src/main/java/org/checkerframework/checker/objectconstruction/ObjectConstructionTransfer.java:235]: Is this method copy-pasted from somewhere else?

File: object-construction-checker/src/main/java/org/checkerframework/checker/objectconstruction/ObjectConstructionTransfer.java [hunk 10]
```diff
@@ -167,9 +227,41 @@ public void accumulate(Node node, TransferResult<CFValue, CFStore> result, Strin
...
+            uniqueName("tmp"),
+            ee.getEnclosingElement(),
+            (ExpressionTree) method.getTree());
+    return iteratorVariable;
+  }
+
+  protected long uid = 0;
+
+  protected String uniqueName(String prefix) {
+    return prefix + "#num" + uid++;
```
Comment (msridhar) [inline object-construction-checker/src/main/java/org/checkerframework/checker/objectconstruction/ObjectConstructionTransfer.java:264]: This code seems to be copy-pasted based on local variable names.  It may be hard to share the logic, but at least rename the variables to be appropriate to this usage.

File: object-construction-checker/tests/mustcall/ACMethodInvocationTest.java [hunk 11]
```diff
@@ -32,17 +32,11 @@ void c() {}
         return f;
     }
 
-    //TODO
     void CallMethodsInSequence() {
-        // won't pass yet due to dataflow issue
-        // :: error: required.method.not.called
```
Comment (msridhar) [inline object-construction-checker/tests/mustcall/ACMethodInvocationTest.java:38]: 🎉 for both these tests passing!

File: object-construction-checker/tests/socket/ACSocketTest.java [hunk 12]
```diff
@@ -214,19 +214,6 @@ void replaceVarWithNull(String address, int port) {
 
     }
 
-//    @EnsuresCalledMethodsIf(expression = "#1", methods = {"close"}, result = true)
```
Comment (msridhar) [inline object-construction-checker/tests/socket/ACSocketTest.java:217]: why is this deleted?

File: object-construction-checker/tests/socket/ACSocketTest.java [hunk 13]
```diff
@@ -333,72 +320,6 @@ void makePrependableSocket() throws IOException {
         final PrependableSocket prependableSocket = new PrependableSocket(null);
     }
 
-
-
-
-
-//    private void acceptConnections() {
```
Comment (msridhar) [inline object-construction-checker/tests/socket/ACSocketTest.java:340]: why is all this commented out?

File: unconnected-socket-checker/src/main/java/org/checkerframework/checker/unconnectedsocket/UnconnectedSocketTransfer.java [hunk 14]
```diff
@@ -7,8 +7,9 @@
 import org.checkerframework.dataflow.cfg.node.FieldAccessNode;
 import org.checkerframework.dataflow.cfg.node.MethodInvocationNode;
 import org.checkerframework.dataflow.cfg.node.Node;
-import org.checkerframework.dataflow.expression.FlowExpressions;
-import org.checkerframework.dataflow.expression.Receiver;
+// import org.checkerframework.dataflow.expression.FlowExpressions;
```
Comment (msridhar) [inline unconnected-socket-checker/src/main/java/org/checkerframework/checker/unconnectedsocket/UnconnectedSocketTransfer.java:10]: we should get rid of these commented imports

File: object-construction-checker/src/main/java/org/checkerframework/checker/objectconstruction/MustCallInvokedChecker.java [hunk 15]
```diff
@@ -219,16 +219,50 @@ private void checkPseudoAssignToOwning(Node node) {
...
-      AnnotationMirror cmAnno =
-          typeFactory.getAnnotatedType(tree).getAnnotationInHierarchy(typeFactory.top);
-      if (!calledMethodsSatisfyMustCall(mustCallVal, cmAnno)) {
-        checker.reportError(
-            tree,
-            "required.method.not.called",
-            MustCallInvokedChecker.formatMissingMustCallMethods(mustCallVal),
-            TreeUtils.typeOf(tree).toString(),
-            "never assigned to an @Owning location");
+      if (typeFactory.biMap.inverse().containsKey(node)) {
```
Comment (msridhar) [inline object-construction-checker/src/main/java/org/checkerframework/checker/objectconstruction/MustCallInvokedChecker.java:222]: This code definitely needs some comments; I can't immediately see what it is doing.  Also, I was imagining that this whole `checkPseudoAssignToOwning` method would go away once temporary variables are properly implemented.  Do you agree?

**Review by msridhar — COMMENTED** [review]
Some initial comments, mostly around documentation and code cleanness.  Unfortunately CI won't pass until the relevant changes get merged into a Checker Framework release; please confirm that `./gradlew build` passes locally

File: object-construction-checker/src/main/java/org/checkerframework/checker/objectconstruction/ObjectConstructionTransfer.java (hunk 5 above)
Comment (Nargeshdb) [reply object-construction-checker/src/main/java/org/checkerframework/checker/objectconstruction/ObjectConstructionTransfer.java]: That's correct, I updated the name and added Javadoc:
https://github.com/kelloggm/object-construction-checker/pull/291/files#diff-3fc5dcd77eafbbf61a8da8156fac14c917b05d258d829aab8e8706b815b91b52R137

File: object-construction-checker/src/main/java/org/checkerframework/checker/objectconstruction/ObjectConstructionTransfer.java (hunk 7 above)
Comment (Nargeshdb) [reply object-construction-checker/src/main/java/org/checkerframework/checker/objectconstruction/ObjectConstructionTransfer.java]: https://github.com/kelloggm/object-construction-checker/pull/291/files#diff-3fc5dcd77eafbbf61a8da8156fac14c917b05d258d829aab8e8706b815b91b52R148

File: object-construction-checker/src/main/java/org/checkerframework/checker/objectconstruction/ObjectConstructionTransfer.java (hunk 9 above)
Comment (Nargeshdb) [reply object-construction-checker/src/main/java/org/checkerframework/checker/objectconstruction/ObjectConstructionTransfer.java:257]: Yes, It's from AccumulationTransfer class.

File: object-construction-checker/src/main/java/org/checkerframework/checker/objectconstruction/ObjectConstructionTransfer.java (hunk 10 above)
Comment (Nargeshdb) [reply object-construction-checker/src/main/java/org/checkerframework/checker/objectconstruction/ObjectConstructionTransfer.java]: I renamed the variables to "temp-var-id"

File: build.gradle [hunk 16]
```diff
@@ -12,7 +12,7 @@ plugins {
 
 
 ext.versions = [
-    checkerFramework: "3.8.0",
+    checkerFramework: "3.9.2-SNAPSHOT",
```
Comment (msridhar) [inline build.gradle:15]: Make this `3.10.0` I think

File: object-construction-checker/src/main/java/org/checkerframework/checker/objectconstruction/ObjectConstructionAnnotatedTypeFactory.java [hunk 17]
```diff
@@ -31,6 +34,11 @@
  */
 public class ObjectConstructionAnnotatedTypeFactory extends CalledMethodsAnnotatedTypeFactory {
 
+  /**
+   * Bidirectional map to preserve temporal variables created for nodes with non-empty @MustCall
+   * annotation and the corresponding nodes
+   */
+  protected BiMap<LocalVariableNode, Tree> mapTempVarToNode = HashBiMap.create();
```
Comment (msridhar) [inline object-construction-checker/src/main/java/org/checkerframework/checker/objectconstruction/ObjectConstructionAnnotatedTypeFactory.java:41]: Just `tempVarToNode`:
```suggestion
  protected BiMap<LocalVariableNode, Tree> tempVarToNode = HashBiMap.create();
```

File: object-construction-checker/src/main/java/org/checkerframework/checker/objectconstruction/ObjectConstructionAnnotatedTypeFactory.java (hunk 2 above)
Comment (msridhar) [reply object-construction-checker/src/main/java/org/checkerframework/checker/objectconstruction/ObjectConstructionAnnotatedTypeFactory.java:85]: I still have this question

File: object-construction-checker/src/main/java/org/checkerframework/checker/objectconstruction/ObjectConstructionAnnotatedTypeFactory.java [hunk 18]
```diff
@@ -87,6 +99,9 @@ public void postAnalyze(ControlFlowGraph cfg) {
         getTypeFactoryOfSubchecker(MustCallChecker.class);
     AnnotatedTypeMirror mustCallAnnotatedType =
         mustCallAnnotatedTypeFactory.getAnnotatedType(element);
+    if (mustCallAnnotatedTypeFactory == null) {
```
Comment (msridhar) [inline object-construction-checker/src/main/java/org/checkerframework/checker/objectconstruction/ObjectConstructionAnnotatedTypeFactory.java:102]: How is this change related to temporary variables?

File: object-construction-checker/src/main/java/org/checkerframework/checker/objectconstruction/ObjectConstructionTransfer.java [hunk 19]
```diff
@@ -45,6 +57,9 @@
 public class ObjectConstructionTransfer extends CalledMethodsTransfer {
   private final ObjectConstructionAnnotatedTypeFactory atypefactory;
 
+  /** TreeBuilder for building new AST nodes */
```
Comment (msridhar) [inline object-construction-checker/src/main/java/org/checkerframework/checker/objectconstruction/ObjectConstructionTransfer.java:60]:
```suggestion
  /** TreeBuilder for building new AST nodes, for creation of temporary variables */
```

File: object-construction-checker/src/main/java/org/checkerframework/checker/objectconstruction/ObjectConstructionTransfer.java [hunk 20]
```diff
@@ -58,6 +73,16 @@
...
     super(analysis);
     this.atypefactory = (ObjectConstructionAnnotatedTypeFactory) analysis.getTypeFactory();
+    ProcessingEnvironment env = atypeFactory.getChecker().getProcessingEnvironment();
+    treeBuilder = new TreeBuilder(env);
+  }
+
+  @Override
+  public TransferResult<CFValue, CFStore> visitTernaryExpression(
+      TernaryExpressionNode node, TransferInput<CFValue, CFStore> input) {
+    TransferResult<CFValue, CFStore> result = super.visitTernaryExpression(node, input);
```
Comment (msridhar) [inline object-construction-checker/src/main/java/org/checkerframework/checker/objectconstruction/ObjectConstructionTransfer.java:83]: Not exactly related to this code, but did we add a test for calling `close()` on a ternary expression directly?

File: object-construction-checker/src/main/java/org/checkerframework/checker/objectconstruction/ObjectConstructionTransfer.java [hunk 21]
```diff
@@ -74,9 +99,65 @@ public ObjectConstructionTransfer(final CFAnalysis analysis) {
             result.getElseStore(),
             exceptionalStores);
     exceptionalStores = null;
+
+    updateStoreWithTempVar(result, node);
+
+    // If the receiver exists in the mapTempVarToNode, then its temporal variable's type will be
+    // updated.
```
Comment (msridhar) [inline object-construction-checker/src/main/java/org/checkerframework/checker/objectconstruction/ObjectConstructionTransfer.java:106]:
```suggestion
    // If there is a temporary variable for the receiver, update its type.
```

File: object-construction-checker/src/main/java/org/checkerframework/checker/objectconstruction/ObjectConstructionTransfer.java (hunk 9 above)
Comment (msridhar) [reply object-construction-checker/src/main/java/org/checkerframework/checker/objectconstruction/ObjectConstructionTransfer.java:257]: can we find a way to not do copy-paste?

File: object-construction-checker/src/main/java/org/checkerframework/checker/objectconstruction/ObjectConstructionTransfer.java [hunk 22]
```diff
@@ -167,9 +249,49 @@ public void accumulate(Node node, TransferResult<CFValue, CFStore> result, Strin
...
+      CFStore store = result.getRegularStore();
+      store.insertValue(target, newAnno);
     }
-    AnnotationMirror newAnno = atypefactory.createAccumulatorAnnotation(valuesAsList);
-    exceptionalStores.values().stream().forEach(s -> s.insertValue(target, newAnno));
+  }
+
+  protected VariableTree createTemporaryVar(Node method) {
+    ExpressionTree tree = (ExpressionTree) method.getTree();
+    TypeMirror iteratorType = TreeUtils.typeOf(tree);
```
Comment (msridhar) [inline object-construction-checker/src/main/java/org/checkerframework/checker/objectconstruction/ObjectConstructionTransfer.java:272]:
```suggestion
    TypeMirror treeType = TreeUtils.typeOf(tree);
```

File: object-construction-checker/src/main/java/org/checkerframework/checker/objectconstruction/ObjectConstructionTransfer.java [hunk 23]
```diff
@@ -167,9 +249,49 @@ public void accumulate(Node node, TransferResult<CFValue, CFStore> result, Strin
...
+    Element enclosingElement;
+    TreePath path = atypefactory.getPath(tree);
+    if (path == null) {
+      enclosingElement = TreeUtils.elementFromTree(tree).getEnclosingElement();
+    } else {
+      Tree methodTree = TreePathUtil.enclosingMethod(path);
+      enclosingElement = TreeUtils.elementFromTree(methodTree);
+    }
+    // Declare and initialize a new, unique iterator variable
+    VariableTree iteratorVariable =
```
Comment (msridhar) [inline object-construction-checker/src/main/java/org/checkerframework/checker/objectconstruction/ObjectConstructionTransfer.java:282]:
```suggestion
    VariableTree tmpVarTree =
```

File: object-construction-checker/src/main/java/org/checkerframework/checker/objectconstruction/MustCallInvokedChecker.java [hunk 24]
```diff
@@ -690,6 +586,26 @@ private void handleSuccessorBlocks(
             setAssign.stream()
                 .allMatch(assign -> succRegularStore.getValue(assign.localVar) == null);
         if (succ instanceof SpecialBlockImpl || noSuccInfo) {
+
+          // Remove the temporary variable defined for a node that throws an exception from the
+          // exceptional successors
+          if (succAndExcType.second != null) {
+            Node exceptionalNode = removeCasts(((ExceptionBlockImpl) block).getNode());
```
Comment (msridhar) [inline object-construction-checker/src/main/java/org/checkerframework/checker/objectconstruction/MustCallInvokedChecker.java:593]:
```suggestion
            Node exceptionalNode = removeCasts(((ExceptionBlock) block).getNode());
```

File: object-construction-checker/src/main/java/org/checkerframework/checker/objectconstruction/MustCallInvokedChecker.java [hunk 25]
```diff
@@ -123,129 +119,89 @@ void checkMustCallInvoked(ControlFlowGraph cfg) {
           handleReturn((ReturnNode) node, cfg, newDefs);
         } else if (node instanceof MethodInvocationNode || node instanceof ObjectCreationNode) {
           handleInvocation(newDefs, node);
-        } else if (node instanceof TypeCastNode || node instanceof TernaryExpressionNode) {
```
Comment (msridhar) [inline object-construction-checker/src/main/java/org/checkerframework/checker/objectconstruction/MustCallInvokedChecker.java:126]: Why don't we need to do anything for `TypeCastNode` anymore?

File: object-construction-checker/src/main/java/org/checkerframework/checker/objectconstruction/MustCallInvokedChecker.java [hunk 26]
```diff
@@ -123,129 +119,89 @@ void checkMustCallInvoked(ControlFlowGraph cfg) {
...
-    }
-    if (!nestedInCastOrTernary(node) && hasNestedPseudoAssignInvoke(node, defs)) {
-      checkPseudoAssignToOwning(node);
+  private void handleTernary(Node node, Set<ImmutableSet<LocalVarWithTree>> defs) {
+    LocalVariableNode ternaryLocal = typeFactory.mapTempVarToNode.inverse().get(node.getTree());
+
+    Node operand = removeCasts(((TernaryExpressionNode) node).getThenOperand());
+    LocalVariableNode operandLocal = typeFactory.mapTempVarToNode.inverse().get(operand.getTree());
+    if (operandLocal == null || !isVarInDefs(defs, operandLocal)) {
+      operand = removeCasts(((TernaryExpressionNode) node).getElseOperand());
```
Comment (msridhar) [inline object-construction-checker/src/main/java/org/checkerframework/checker/objectconstruction/MustCallInvokedChecker.java:137]: It seems like we never look at the else operand if we have a  local for the then operand.  That's not what I expected; I thought we'd handle them separately, treating each as a separate assignment to the temporary for the ternary expression.  Did I misunderstand?

File: object-construction-checker/src/main/java/org/checkerframework/checker/objectconstruction/MustCallInvokedChecker.java [hunk 27]
```diff
@@ -123,129 +119,89 @@ void checkMustCallInvoked(ControlFlowGraph cfg) {
...
+    updateDefsWithTempVar(defs, node);
   }
 
-  /**
-   * Given a node representing a method or constructor call, checks that if the call has a non-empty
-   * {@code @MustCall} type, then its result is pseudo-assigned to some location that can take
-   * ownership of the result
-   */
-  private void checkPseudoAssignToOwning(Node node) {
+  private void updateDefsWithTempVar(Set<ImmutableSet<LocalVarWithTree>> defs, Node node) {
```
Comment (msridhar) [inline object-construction-checker/src/main/java/org/checkerframework/checker/objectconstruction/MustCallInvokedChecker.java:165]: This method needs comments.  I can't follow the logic.

File: object-construction-checker/src/main/java/org/checkerframework/checker/objectconstruction/MustCallInvokedChecker.java [hunk 28]
```diff
@@ -255,32 +211,17 @@ private boolean isOwningAssignmentLhs(Element elem) {
    * a super constructor call, the method's return type is annotated {@link NotOwning}, or the
    * receiver parameter is in the {@code defs} and has @MustCallChoice annotation.
    */
-  private boolean shouldSkipInvokePseudoAssignCheck(
-      Node node, Set<ImmutableSet<LocalVarWithTree>> defs) {
+  private boolean shouldSkipInvokePseudoAssignCheck(Node node) {
```
Comment (msridhar) [inline object-construction-checker/src/main/java/org/checkerframework/checker/objectconstruction/MustCallInvokedChecker.java:214]: rename this method?  since we're not doing pseudoassign checks anymore?

File: object-construction-checker/tests/socket/ACSocketTest.java (hunk 13 above)
Comment (msridhar) [reply object-construction-checker/tests/socket/ACSocketTest.java:340]: still have the same question

File: object-construction-checker/tests/socket/GetChannelOnLocks.java [hunk 29]
```diff
@@ -0,0 +1,36 @@
+import java.io.IOException;
+import java.nio.channels.FileLock;
+
+class GetChannelOnLocks {
```
Comment (msridhar) [inline object-construction-checker/tests/socket/GetChannelOnLocks.java:4]: Add comments on what we are testing for here

File: object-construction-checker/src/main/java/org/checkerframework/checker/objectconstruction/MustCallInvokedChecker.java [hunk 30]
```diff
@@ -123,129 +119,89 @@ void checkMustCallInvoked(ControlFlowGraph cfg) {
...
-                        NotOwning.class)
-                    == null
-                : typeFactory.getDeclAnnotation(elementForType, Owning.class) != null;
-      } else if (assignmentContext instanceof LambdaReturnContext) {
-        // TODO handle this case.  For now we will report an error
-      } else {
-        throw new BugInCF("unexpected AssignmentContext type " + assignmentContext.getClass());
+    if (typeFactory.mapTempVarToNode.inverse().containsKey(node.getTree())) {
+
+      LocalVariableNode temporaryLocal = typeFactory.mapTempVarToNode.inverse().get(node.getTree());
```
Comment (msridhar) [inline object-construction-checker/src/main/java/org/checkerframework/checker/objectconstruction/MustCallInvokedChecker.java:169]: Move this line before the `if` check, and change the `if` check to see if `temporaryLocal == null`

File: object-construction-checker/src/main/java/org/checkerframework/checker/objectconstruction/MustCallInvokedChecker.java [hunk 31]
```diff
@@ -123,129 +119,89 @@ void checkMustCallInvoked(ControlFlowGraph cfg) {
...
-            MustCallInvokedChecker.formatMissingMustCallMethods(mustCallVal),
-            TreeUtils.typeOf(tree).toString(),
-            "never assigned to an @Owning location");
+
+      if (receiver == null
+          && node instanceof MethodInvocationNode
+          && (typeFactory.returnsThis((MethodInvocationTree) tree))) {
+        receiver = ((MethodInvocationNode) node).getTarget().getReceiver();
+        if (receiver instanceof MethodInvocationNode) {
+          receiver = typeFactory.mapTempVarToNode.inverse().get(receiver.getTree());
```
Comment (msridhar) [inline object-construction-checker/src/main/java/org/checkerframework/checker/objectconstruction/MustCallInvokedChecker.java:183]: We should add a method `getTempVarForTree()` to the type factory that encapsulates this logic of doing a `get()` on the `inverse()`, and use that everywhere

File: object-construction-checker/src/main/java/org/checkerframework/checker/objectconstruction/MustCallInvokedChecker.java [hunk 32]
```diff
@@ -123,129 +119,89 @@ void checkMustCallInvoked(ControlFlowGraph cfg) {
...
-        // TODO handle this case.  For now we will report an error
-      } else {
-        throw new BugInCF("unexpected AssignmentContext type " + assignmentContext.getClass());
+    if (typeFactory.mapTempVarToNode.inverse().containsKey(node.getTree())) {
+
+      LocalVariableNode temporaryLocal = typeFactory.mapTempVarToNode.inverse().get(node.getTree());
+      LocalVarWithTree lhsLocalVarWithTreeNew =
+          new LocalVarWithTree(new LocalVariable(temporaryLocal), tree);
+
+      Node receiver = null;
```
Comment (msridhar) [inline object-construction-checker/src/main/java/org/checkerframework/checker/objectconstruction/MustCallInvokedChecker.java:173]: It's here that I start to get lost on what the method is doing

File: object-construction-checker/src/main/java/org/checkerframework/checker/objectconstruction/MustCallInvokedChecker.java [hunk 33]
```diff
@@ -445,107 +343,100 @@ private Node removeCasts(Node node) {
   }
 
   private void handleAssignFromRHS(
-      AssignmentNode node, Set<ImmutableSet<LocalVarWithTree>> newDefs, Node rhs) {
+      AssignmentNode node, Set<ImmutableSet<LocalVarWithTree>> newDefs, Node rhstemp) {
     Node lhs = node.getTarget();
     Element lhsElement = TreeUtils.elementFromTree(lhs.getTree());
+    if (typeFactory.mapTempVarToNode.inverse().containsKey(rhstemp.getTree())) {
```
Comment (msridhar) [inline object-construction-checker/src/main/java/org/checkerframework/checker/objectconstruction/MustCallInvokedChecker.java:349]: This logic is weird.  Why did we rename the parameter to `rhstemp`, which is kind of strange?  Can we write this code so it doesn't re-assign the parameter?

File: object-construction-checker/src/main/java/org/checkerframework/checker/objectconstruction/MustCallInvokedChecker.java [hunk 34]
```diff
@@ -724,6 +640,10 @@ private void handleSuccessorBlocks(
     }
   }
 
+  private boolean isTempVar(LocalVariable localVariable) {
+    return localVariable.getElement().getSimpleName().toString().contains("temp-var");
```
Comment (msridhar) [inline object-construction-checker/src/main/java/org/checkerframework/checker/objectconstruction/MustCallInvokedChecker.java:644]: can't we add a method to the type factory to check this properly by checking for membership in the bi-map?

**Review by msridhar — CHANGES_REQUESTED** [review]
Looking promising!  I have a few comments

**Comment by msridhar** [comment]
Another thought.  Would this code become even simpler if we created temporary variables for downcasts as well?  There are various places where we unwrap downcasts that I think could be simplified.

File: build.gradle (hunk 16 above)
Comment (Nargeshdb) [reply build.gradle]: https://github.com/kelloggm/object-construction-checker/pull/291/commits/790ac07e99314570d32d4c8b2fc03158226ec15a

File: object-construction-checker/src/main/java/org/checkerframework/checker/objectconstruction/ObjectConstructionAnnotatedTypeFactory.java (hunk 17 above)
Comment (Nargeshdb) [reply object-construction-checker/src/main/java/org/checkerframework/checker/objectconstruction/ObjectConstructionAnnotatedTypeFactory.java]: https://github.com/kelloggm/object-construction-checker/pull/291/commits/790ac07e99314570d32d4c8b2fc03158226ec15a

File: object-construction-checker/src/main/java/org/checkerframework/checker/objectconstruction/ObjectConstructionAnnotatedTypeFactory.java (hunk 2 above)
Comment (Nargeshdb) [reply object-construction-checker/src/main/java/org/checkerframework/checker/objectconstruction/ObjectConstructionAnnotatedTypeFactory.java:85]: I added this check to handle `NullPointerException` [here](https://github.com/kelloggm/object-construction-checker/blob/temporaryVar/object-construction-checker/src/main/java/org/checkerframework/checker/objectconstruction/ObjectConstructionTransfer.java#L103) when the given `node` is `(this).<init>()`. I got a lot of `NullPointerException` in AutoValue test cases.

File: object-construction-checker/src/main/java/org/checkerframework/checker/objectconstruction/ObjectConstructionTransfer.java (hunk 9 above)
Comment (Nargeshdb) [reply object-construction-checker/src/main/java/org/checkerframework/checker/objectconstruction/ObjectConstructionTransfer.java:255]: Can we make this method public in AccumulationTransfer?

File: object-construction-checker/tests/socket/ACSocketTest.java (hunk 12 above)
Comment (Nargeshdb) [reply object-construction-checker/tests/socket/ACSocketTest.java]: I thought it's repetitive but it's not, so I added it again.
https://github.com/kelloggm/object-construction-checker/pull/291/commits/32c6c64d3fe2e77413bc1fadc0878fd07aea1cfd

File: object-construction-checker/tests/socket/ACSocketTest.java (hunk 13 above)
Comment (Nargeshdb) [reply object-construction-checker/tests/socket/ACSocketTest.java]: It was commented out, I just removed it because it contains some methods/fields that are not defined in this class, so we can't test it. I added it again https://github.com/kelloggm/object-construction-checker/pull/291/commits/03e52aad72ffba8808e8d6ef71189925d887c288

File: object-construction-checker/src/main/java/org/checkerframework/checker/objectconstruction/ObjectConstructionAnnotatedTypeFactory.java (hunk 18 above)
Comment (Nargeshdb) [reply object-construction-checker/src/main/java/org/checkerframework/checker/objectconstruction/ObjectConstructionAnnotatedTypeFactory.java:102]: https://github.com/kelloggm/object-construction-checker/pull/291/files/424f947ae09c7edf540ad6c5c9ad821549ab31ed#r570794811

File: object-construction-checker/src/main/java/org/checkerframework/checker/objectconstruction/MustCallInvokedChecker.java (hunk 34 above)
Comment (Nargeshdb) [reply object-construction-checker/src/main/java/org/checkerframework/checker/objectconstruction/MustCallInvokedChecker.java:643]: I removed this method because I don't use it.

File: object-construction-checker/src/main/java/org/checkerframework/checker/objectconstruction/ObjectConstructionTransfer.java (hunk 20 above)
Comment (Nargeshdb) [reply object-construction-checker/src/main/java/org/checkerframework/checker/objectconstruction/ObjectConstructionTransfer.java:83]: https://github.com/kelloggm/object-construction-checker/pull/291/files#diff-d9c49afa35f8d4df342e326e80fd73fd3d5df0b9c0d3a1c95cd6662f84251d00R356

File: object-construction-checker/src/main/java/org/checkerframework/checker/objectconstruction/ObjectConstructionTransfer.java (hunk 9 above)
Comment (msridhar) [reply object-construction-checker/src/main/java/org/checkerframework/checker/objectconstruction/ObjectConstructionTransfer.java:254]: @Nargeshdb we can make it `protected` and then call it from here

**Review by msridhar — CHANGES_REQUESTED** [review]
Just one minor thing.  We need to remember as follow-ups to (1) fix the known false positives from weird ternary expressions and (2) also introduce temporary variables for downcasts

File: object-construction-checker/src/main/java/org/checkerframework/checker/objectconstruction/ObjectConstructionTransfer.java (hunk 9 above)
Comment (msridhar) [reply object-construction-checker/src/main/java/org/checkerframework/checker/objectconstruction/ObjectConstructionTransfer.java:254]: This can be a follow up change as it requires a change in CF
